### ISSUE TITLE
How to use the PPO and SAC algorithms from a file to generate expert trajectories?

### ISSUE THREAD

**Issue by zichengS1999** [bAmpT/mgail-pytorch#1]
Hello, I would like to use these algorithms on a reinforcement learning environment I've built, but I haven't found the part about generating expert trajectories. If you have the time, please guide me through it.Thank you very much!

**Comment by bAmpT** [comment]
Hi, in order to generate expert trajectories you can use checkpoints from the SAC or PPO algorithms and then record trajectories in the buffer memory.

You're right that those were missing, I've now added the scripts I used for training and creating the buffer memory.

See here: [collect_buffer.py](https://github.com/bAmpT/mgail-pytorch/blob/master/expert/collect_buffer.py), [train_expert.py](https://github.com/bAmpT/mgail-pytorch/blob/master/expert/train_expert.py)

**Comment by zichengS1999** [comment]
"Thank you very much for your reply; it has been very helpful to me! Also, I have another question: What is the content of the agent_trajectories folder? Is it the random interaction of the agent in the reinforcement learning environment, or is it interaction with a specific policy, or something else? What is the specific number of interactions recorded in it?"

**Comment by zichengS1999** [comment]
"When I started training, the issue of gradient vanishing occurred during the training process of the forward model. The training was interrupted in the debugger (pdb), and p.grad was found to be None in the mentioned statement."

**Comment by bAmpT** [comment]
> "Thank you very much for your reply; it has been very helpful to me! Also, I have another question: What is the content of the agent_trajectories folder? Is it the random interaction of the agent in the reinforcement learning environment, or is it interaction with a specific policy, or something else? What is the specific number of interactions recorded in it?"

The agent trajectories are the interactions of the agent in the rl environment during training, the number of trajectories is a hyper parameter that can be set in the env.er_agent_size.


> "When I started training, the issue of gradient vanishing occurred during the training process of the forward model. The training was interrupted in the debugger (pdb), and p.grad was found to be None in the mentioned statement."

Depending on the forward model you are using vanishing gradients can be a problem. I did some experiments with a VAE model, during training it is a common problem that you can get some NaN values. I tried to get rid of them with clipping the gradient norm, but you can try to use a different initialization and/or other forms of normalization.

The forward model with a VAE is inspired by the PPUU paper ([https://github.com/Atcold/pytorch-PPUU](https://github.com/Atcold/pytorch-PPUU))

**Comment by zichengS1999** [comment]
> > "Thank you very much for your reply; it has been very helpful to me! Also, I have another question: What is the content of the agent_trajectories folder? Is it the random interaction of the agent in the reinforcement learning environment, or is it interaction with a specific policy, or something else? What is the specific number of interactions recorded in it?"
> 
> The agent trajectories are the interactions of the agent in the rl environment during training, the number of trajectories is a hyper parameter that can be set in the env.er_agent_size.
> 
> > "When I started training, the issue of gradient vanishing occurred during the training process of the forward model. The training was interrupted in the debugger (pdb), and p.grad was found to be None in the mentioned statement."
> 
> Depending on the forward model you are using vanishing gradients can be a problem. I did some experiments with a VAE model, during training it is a common problem that you can get some NaN values. I tried to get rid of them with clipping the gradient norm, but you can try to use a different initialization and/or other forms of normalization.
> 
> The forward model with a VAE is inspired by the PPUU paper (https://github.com/Atcold/pytorch-PPUU)

Thank you for your response. I have resolved the previous issue! I also want to ask about the agent_trajectories file - should it be initialized with a .bin file before training? Should I pre-collect some trajectories through a specific policy, perhaps to better explore the environment?

**Comment by zichengS1999** [comment]
I also want to ask about the issue of the number of expert trajectories. If there is only one trajectory in the environment that completes the task, would providing 1000 trajectories versus 100 or 10 trajectories have an impact on training, such as the issue of overfitting?

**Comment by bAmpT** [comment]
The agent trajectories are usually collected during training, but you can experiment with to use random policy to initialize the buffer. For the expert trajectories it's very much dependent on the complexity of the task you are training...
